File: portfolio-assistant/src/lib/summarizer.ts
/**
 * lib/summarizer.ts
 * -----------------
 * Background summarisation job for long conversations.
 *
 * When trimming occurs, the messages that were cut are lost from the LLM's
 * context window. This module produces a 3–5 sentence rolling summary of
 * those dropped messages and writes it back to the DB, so future requests
 * can inject it as "[Earlier context: ...]".
 *
 * Key design rules:
 *   - Never await this from the request path — fire-and-forget only.
 *   - Never regenerate the full conversation. Always extend from the existing
 *     summary, processing only the newly-trimmed messages.
 *   - Only run if newestTrimmedMessageId !== summarisedUpToId (i.e. the
 *     summary is stale) AND more than MIN_DROPPED_TO_SUMMARISE were cut.
 */

import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";
import { ANTHROPIC_API_KEY, OPENAI_API_KEY, RAG_BACKEND_URL } from "./config";
import type { MessageRead } from "./conversations";

// ── Config ────────────────────────────────────────────────────────────────────

/** Only summarise if at least this many messages were trimmed. */
const MIN_DROPPED_TO_SUMMARISE = 5;

const SUMMARISER_MODELS = {
    anthropic: "claude-haiku-4-5-20251001",
    openai: "gpt-4o-mini",
} as const;

// ── Types ─────────────────────────────────────────────────────────────────────

export interface SummaryTrigger {
    convId: string;
    droppedMessages: MessageRead[];
    newestTrimmedMessageId: string;
    existingSummary: string | null;
    summarisedUpToId: string | null;
}

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Fire-and-forget — call this after streaming finishes.
 * Returns immediately; summarisation runs in the background.
 */
export function maybeTriggerSummarization(trigger: SummaryTrigger): void {
    if (trigger.droppedMessages.length <= MIN_DROPPED_TO_SUMMARISE) return;
    if (trigger.newestTrimmedMessageId === trigger.summarisedUpToId) return;

    // Intentionally not awaited — runs after the response is already sent
    runSummarization(trigger).catch((err) =>
        console.warn("[summarizer] Background job failed:", err),
    );
}

// ── Implementation ────────────────────────────────────────────────────────────

async function runSummarization(trigger: SummaryTrigger): Promise<void> {
    const { convId, droppedMessages, newestTrimmedMessageId, existingSummary } = trigger;

    const prompt = buildPrompt(droppedMessages, existingSummary);

    const newSummary = ANTHROPIC_API_KEY
        ? await summariseWithAnthropic(prompt)
        : await summariseWithOpenAI(prompt);

    await persistSummary(convId, newSummary, newestTrimmedMessageId);
    console.log(`[summarizer] Updated summary for conversation ${convId}`);
}

/**
 * Builds the summarisation prompt.
 * Always extends from the existing summary — never starts from scratch.
 */
function buildPrompt(dropped: MessageRead[], existingSummary: string | null): string {
    const transcript = dropped
        .map((m) => `${m.role.toUpperCase()}: ${m.content}`)
        .join("\n\n");

    const seedSection = existingSummary
        ? `Existing summary (do not repeat this verbatim — extend it):\n${existingSummary}\n\n`
        : "";

    return (
        `${seedSection}` +
        `New messages to incorporate:\n\n${transcript}\n\n` +
        `Write an updated 3–5 sentence summary of this conversation so far. ` +
        `Cover: the user's goal, any documents generated (type + key decisions), ` +
        `preferences stated, and important facts established. ` +
        `Write in present tense. Reply with only the summary text.`
    );
}

async function summariseWithAnthropic(prompt: string): Promise<string> {
    const client = new Anthropic({ apiKey: ANTHROPIC_API_KEY });
    const response = await client.messages.create({
        model: SUMMARISER_MODELS.anthropic,
        max_tokens: 256,
        messages: [{ role: "user", content: prompt }],
    });
    return response.content
        .filter((b) => b.type === "text")
        .map((b) => b.text)
        .join("")
        .trim();
}

async function summariseWithOpenAI(prompt: string): Promise<string> {
    const client = new OpenAI({ apiKey: OPENAI_API_KEY });
    const response = await client.chat.completions.create({
        model: SUMMARISER_MODELS.openai,
        max_tokens: 256,
        messages: [{ role: "user", content: prompt }],
    });
    return (response.choices[0]?.message?.content ?? "").trim();
}

/** Writes the new summary back to the Python backend via the Next.js API proxy. */
async function persistSummary(
    convId: string,
    summary: string,
    summarisedUpToMessageId: string,
): Promise<void> {
    await fetch(`${RAG_BACKEND_URL}/api/v1/conversations/${convId}/summary`, {
        method: "PATCH",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ summary, summarised_up_to_message_id: summarisedUpToMessageId }),
    });
}

File: memra-api/src/memra/domain/services/summarizer.py
"""
core/summarizer.py
------------------
Background summarisation job for long conversations.

Ports lib/summarizer.ts verbatim.

When trimming occurs, the messages that were cut are lost from the LLM's
context window. This module produces a 3-5 sentence rolling summary of
those dropped messages and writes it back to the DB, so future requests
can inject it as "[Earlier context: ...]".

Key design rules:
  - Never await this from the request path — fire-and-forget only.
  - Never regenerate the full conversation. Always extend from the existing
    summary, processing only the newly-trimmed messages.
  - Only run if newest_trimmed_id != summarised_up_to_id (i.e. the
    summary is stale) AND more than MIN_DROPPED_TO_SUMMARISE were cut.
"""

from __future__ import annotations

import asyncio
import logging
from uuid import UUID

logger = logging.getLogger(__name__)

MIN_DROPPED_TO_SUMMARISE = 5


def maybe_trigger_summarization(
    conv_id: str,
    dropped_messages: list,
    newest_trimmed_id: str | None,
    existing_summary: str | None,
    summarised_up_to_id: str | None,
    settings,
    db_session_factory,
) -> None:
    """
    Fire-and-forget — call this after streaming finishes.
    Returns immediately; summarisation runs in the background.
    """
    if not newest_trimmed_id:
        return
    if len(dropped_messages) <= MIN_DROPPED_TO_SUMMARISE:
        return
    if newest_trimmed_id == summarised_up_to_id:
        return

    task = asyncio.create_task(
        _run_summarization(
            conv_id, dropped_messages, newest_trimmed_id,
            existing_summary, settings, db_session_factory,
        )
    )
    task.add_done_callback(
        lambda t: (
            logger.warning("[summarizer] Background job failed: %s", t.exception())
            if not t.cancelled() and t.exception()
            else None
        )
    )


async def _run_summarization(
    conv_id: str,
    dropped_messages: list,
    newest_trimmed_id: str,
    existing_summary: str | None,
    settings,
    db_session_factory,
) -> None:
    prompt = _build_prompt(dropped_messages, existing_summary)

    if settings.anthropic_api_key:
        new_summary, _ = await _summarise_with_anthropic(
            prompt, settings.anthropic_api_key, settings.summarizer_anthropic_model
        )
    else:
        new_summary, _ = await _summarise_with_openai(
            prompt, settings.openai_api_key, settings.summarizer_openai_model
        )

    # Persist directly via service — no HTTP round-trip needed
    from memra.domain.services import conversations as conv_svc
    async with db_session_factory() as session:
        await conv_svc.update_summary(
            session,
            UUID(conv_id),
            new_summary,
            UUID(newest_trimmed_id),
        )
    logger.info("[summarizer] Updated summary for conversation %s", conv_id)


def _build_prompt(dropped: list, existing_summary: str | None) -> str:
    """
    Builds the summarisation prompt.
    Always extends from the existing summary — never starts from scratch.
    """
    transcript = "\n\n".join(f"{m.role.upper()}: {m.content}" for m in dropped)
    seed_section = (
        f"Existing summary (do not repeat this verbatim — extend it):\n{existing_summary}\n\n"
        if existing_summary else ""
    )
    return (
        f"{seed_section}"
        f"New messages to incorporate:\n\n{transcript}\n\n"
        "Write an updated 3-5 sentence summary of this conversation so far. "
        "Cover: the user's goal, any documents generated (type + key decisions), "
        "preferences stated, and important facts established. "
        "Write in present tense. Reply with only the summary text."
    )


async def _summarise_with_anthropic(
    prompt: str, api_key: str, model: str
) -> tuple[str, dict]:
    import anthropic
    client = anthropic.AsyncAnthropic(api_key=api_key)
    response = await client.messages.create(
        model=model,
        max_tokens=256,
        messages=[{"role": "user", "content": prompt}],
    )
    text = "".join(b.text for b in response.content if b.type == "text").strip()
    usage = {
        "provider": "anthropic",
        "model": model,
        "input_tokens": response.usage.input_tokens,
        "output_tokens": response.usage.output_tokens,
    }
    return text, usage


async def _summarise_with_openai(
    prompt: str, api_key: str, model: str
) -> tuple[str, dict]:
    from openai import AsyncOpenAI
    client = AsyncOpenAI(api_key=api_key)
    response = await client.chat.completions.create(
        model=model,
        max_tokens=256,
        messages=[{"role": "user", "content": prompt}],
    )
    text = (response.choices[0].message.content or "").strip()
    usage = {
        "provider": "openai",
        "model": model,
        "input_tokens": response.usage.prompt_tokens if response.usage else None,
        "output_tokens": response.usage.completion_tokens if response.usage else None,
    }
    return text, usage
